### ISSUE TITLE
non ascii characters like `ß` `ä` `ü` `ö` (german) or `é` `á` (spanish) break ttyper

### ISSUE THREAD

**Issue by Quantenzitrone** [max-niederman/ttyper#77]
### To reproduce:
- start ttyper with `ttyper -l german`
- type until you encounter a word using `ß` `ä` `ü` or `ö` (they are pretty common)
-> ttyper panics

### Rust backtrace
```
schon können thread 'main' panicked at 'byte index 1 is not a char boundary; it is inside 'ö' (bytes 0..2) of `önnen `', /build/rustc-1.67.1-src/library/core/src/str/mod.rs:678:13r waren denen wurden vor anders einzelnen sind euch Mann war dir allen welcher aus mir wohl                                          │

stack backtrace:
0:     0x55a355c1a799 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::haaf5fb1112692432 
1:     0x55a355c4d24e - core::fmt::write::hc9810906af1f037c
2:     0x55a355c2c945 - std::io::Write::write_fmt::h42570a50d1c05a1d
3:     0x55a355c1a555 - std::sys_common::backtrace::print::hdb4dd2f5727cb1da
4:     0x55a355c39b5f - std::panicking::default_hook::{{closure}}::hbf6435d4555f4c42
5: 0x55a355c39885 - std::panicking::default_hook::hb6b577f7a430b818
6:     0x55a355c3a17f - std::panicking::rust_panic_with_hook::h2dbb792ca6451cd5
7:  0x55a355c1ab09 - std::panicking::begin_panic_handler::{{closure}}::h78627a3718fcdf2c
8:     0x55a355c1a8ec - std::sys_common::backtrace::__rust_end_short_backtrace::h79f584276ca2060b
9:     0x55a355c39d12 - rust_begin_unwind
10:     0x55a355b4ccd3 - core::panicking::panic_fmt::ha005c52a737c94d3
11:     0x55a355c5519f - core::str::slice_error_fail_rt::hee1648429b192c24
12:     0x55a355b4d1d7 - core::str::slice_error_fail::h8d6e6677c0da10d6
13:     0x55a355b882ed - <&ttyper::test::Test as ttyper::ui::ThemedWidget>::render::h438f2d3e6c128fd3
14:     0x55a355b84aa2 - tui::terminal::Terminal<B>::draw::h236e6581cbaba0d3
15:     0x55a355b84278 - ttyper::main::h8df3b922cb9204c0
16:     0x55a355b6de93 - std::sys_common::backtrace::__rust_begin_short_backtrace::he8ad18538a2ba091
17:     0x55a355b85521 - std::rt::lang_start::{{closure}}::h983486b71761f827
18:     0x55a355c2bbd4 - std::rt::lang_start_internal::h62a41a9744fc6640 
19:     0x55a355b847f5 - main
20:     0x7f280642924e - __libc_start_call_main 
21:     0x7f2806429309 - __libc_start_main_alias_1
22:     0x55a355b4d295 - _start
23:                0x0 - <unknown>
```

**Comment by Quantenzitrone** [comment]
same with spanish `á` `é` etc.

**Comment by Quantenzitrone** [comment]
i can't imagine no one encountered this yet...
ttyper panics for literally any language, that is not english nor a programming language that only uses ascii chars.
This must have to do with my setup.
- Terminals: i tested konsole, xterm and the agetty
- Keyboard layout: tested with `bone` and `us` (xorg and tty)
- ttyper version: 1.1.0
- Distro: NixOS

**Comment by max-niederman** [comment]
Yeah I've specifically tested Unicode characters on earlier versions before and I haven't seen this issue. I'll test on my machine later today.

**Comment by leo848** [comment]
Simple repro for QWERTZ keyboard:

```shell
echo "ääääää" > /tmp/wordlist
ttyper --language-file /tmp/wordlist
```

Then, press the <kbd>ä</kbd> key. Expected behavior: one letter turns green, actual behavior: two letters turn green (but you still have to type six to complete the word). 

This might also be reproduceable on other keyboard layouts, but I'm not sure. 

---

I would probably check for instances of `string.len()` where the correct method would actually be `string.chars().count()`, this would be my first instinct to where this bug might lie.

**Comment by leo848** [comment]
The error seems to be in this method:

https://github.com/max-niederman/ttyper/blob/main/src/ui.rs#L79-L184

I already tried changing all of the `String::len` calls with `Chars::count`, but that didn't help. Unicode segmentation might be worth a try fwiw. However, there are several calls to `String::split_at`, which is be the cause of the panic. I changed https://github.com/max-niederman/ttyper/blob/main/src/ui.rs#L131 to not always use the next byte, but the next char boundary - this actually helped a bit, by not panicking instantly at startup, but after one character is typed. https://github.com/max-niederman/ttyper/blob/main/src/ui.rs#L128 seems to be problematic, but I don't see how (as it persists with `Chars::count`).
